### PR TITLE
Adds an experimental version of Model.js

### PR DESCRIPTION
**Pull request by sanjeetsuhag**
This PR adds `ModelExperimental.js` and the required classes to support it. With this new `ModelExperimental` class, the intention is to create a model handling interface that takes full advantage of the `ModelComponents` and `GltfLoader`.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @sanjeetsuhag!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

File: Source/Core/ExperimentalFeatures.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+/**
```
Comment (ptrgags) [inline Source/Core/ExperimentalFeatures.js:1]: I'd like to make this comment more in depth to give some guidance on when it's appropriate to use this, see #9639. I can do this one in my documentation sweep

File: Source/Shaders/ModelExperimental/GeometryVS.glsl [hunk 2]
```diff
@@ -0,0 +1,32 @@
+void processPoints() {
```
Comment (ptrgags) [inline Source/Shaders/ModelExperimental/GeometryVS.glsl:1]: @sanjeetsuhag one thing we both should note is our Coding Guide https://github.com/CesiumGS/cesium/blob/main/Documentation/Contributors/CodingGuide/README.md#formatting-1 uses 4-space tabs and braces on their own line.

File: Source/Scene/ModelExperimental/ModelExperimentalSceneGraph.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+import buildDrawCommand from "./buildDrawCommand.js";
+import defaultValue from "../../Core/defaultValue.js";
+import defined from "../../Core/defined.js";
+import Matrix4 from "../../Core/Matrix4.js";
+import ModelExperimentalSceneMeshPrimitive from "./ModelExperimentalSceneMeshPrimitive.js";
+import ModelExperimentalSceneNode from "./ModelExperimentalSceneNode.js";
+import ModelExperimentalUtility from "./ModelExperimentalUtility.js";
+import RenderResources from "./RenderResources.js";
+
+export default function ModelExperimentalSceneGraph(options) {
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/ModelExperimentalSceneGraph.js:11]: @sanjeetsuhag do you want me to document this in my documentation sweep or do you want to handle this one since you're working on the scene graph?

File: Source/Scene/ModelExperimental/ModelExperimentalSceneMeshPrimitive.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import GeometryPipelineStage from "./GeometryPipelineStage.js";
+
+export default function ModelExperimentalSceneMeshPrimitive(options) {
+  /**
+   * @type {ModelComponents.Primitive}
+   */
+  this._primitive = options.primitive;
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/ModelExperimentalSceneMeshPrimitive.js:7]: We talked about this in slack. We use this class like a simple struct, so I'll update these variables to not use underscores and document the fields properly

File: Source/Scene/ModelExperimental/ModelExperimentalSceneNode.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+export default function ModelSceneNode(options) {
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/ModelExperimentalSceneNode.js:1]: Similarly here, I'll make this more struct-like

File: Source/Scene/ModelExperimental/ModelExperimentalSceneNode.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+  /**
+   * @type {ModelExperimentalSceneMeshPrimitive}
+   */
+  this._sceneMeshPrimitives = [];
+
+  initialize(this);
+}
+
+function initialize(sceneNode) {
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/ModelExperimentalSceneNode.js:25]: @sanjeetsuhag I presume this function is empty because it won't be until instancing that you fill this out?

File: Source/Scene/ModelExperimental/RenderResources.js [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+/**
+ * Resources assigned at each level of a 3D model (model, node, mesh primitive)
+ * that are required to generate {@link DrawCommand}s. These resources are passed
+ * through the various model pipeline stages (see for example
+ * {@link GeometryPipelineStage}) and updated in place. Finally,
+ * {@link buildDrawCommand} is called to construct the draw command.
+ *
+ * @private
+ */
+var RenderResources = {};
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/RenderResources.js:15]: I plan to add some unit tests for these classes to make sure fields are propagated correctly, especially when inheriting fields

**Review by ptrgags — COMMENTED** [review]
@sanjeetsuhag this isn't a full review, just taking some notes about cleanup tasks, especially concerning documentation. I'll take care of most of them this afternoon.

File: Source/Scene/ModelExperimental/ModelExperimentalSceneNode.js (hunk 6 above)
Comment (sanjeetsuhag) [reply Source/Scene/ModelExperimental/ModelExperimentalSceneNode.js:25]: Correct.

File: Source/Scene/ModelExperimental/ModelExperimentalSceneGraph.js (hunk 3 above)
Comment (sanjeetsuhag) [reply Source/Scene/ModelExperimental/ModelExperimentalSceneGraph.js:11]: You can take this one.

File: Source/Scene/ModelExperimental/ModelExperimentalSceneNode.js (hunk 6 above)
Comment (ptrgags) [reply Source/Scene/ModelExperimental/ModelExperimentalSceneNode.js:68]: cool, I put a comment in the code to make this clear

File: Source/Core/ExperimentalFeatures.js (hunk 1 above)
Comment (ptrgags) [reply Source/Core/ExperimentalFeatures.js:1]: @sanjeetsuhag Updated.

@ebogo1 you might also want to look over this comment too to see if this is a good description of the things we discussed in #9639 so we're all on the same page.

File: Source/Scene/ModelExperimental/ModelExperimentalSceneMeshPrimitive.js (hunk 4 above)
Comment (ptrgags) [reply Source/Scene/ModelExperimental/ModelExperimentalSceneMeshPrimitive.js]: I made this change earlier, though I still need to try this out in a sandcastle to make sure I updated other places properly.

**Review by ptrgags — COMMENTED** [review]
@sanjeetsuhag I reviewed most of this PR, there's still a couple more files I want to look at tomorrow morning.

**Comment by sanjeetsuhag** [comment]
I was doing some experimenting and I realized that we need to think about the `destroy` function for `ModelExperimental` for when the primitive is removed from the scene. It needs to be propagated to the render resources, so we can clear everything out.

**Comment by ptrgags** [comment]
Ah the typescript build error is probably my fault, I need to check the use of `@private` in the JSDoc comments.

Agreed, we need to think about the `destroy()` function. We need a way to track any resources we allocate that are not part of the `GltfLoader` (which right now is nothing, but eventually we will have this with instancing, batch textures, etc.). @sanjeetsuhag I think we should open an issue for this, as this goes beyond the scope of this PR for the core functionality.

File: Source/Scene/ModelExperimental/GeometryPipelineStage.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+GeometryPipelineStage.process = function (renderResources, primitive) {
+  var attributeIndex = 0;
+  for (var i = 0; i < primitive.attributes.length; i++) {
+    var attribute = primitive.attributes[i];
+    processAttribute(
+      renderResources,
+      attribute,
+      attribute.semantic === VertexAttributeSemantic.POSITION
+        ? 0
+        : ++attributeIndex
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/GeometryPipelineStage.js:39]: Personally I don't like seeing `++` embedded in larger expressions. I'd put the increment on its own line.

**Review by ptrgags — COMMENTED** [review]
@sanjeetsuhag finished reviewing GeometryPipeline stage, looks solid overall

File: Source/Scene/ModelExperimental/buildDrawCommand.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
+import DrawCommand from "../../Renderer/DrawCommand.js";
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/buildDrawCommand.js:1]: Oh @sanjeetsuhag I just checked the code coverage, looks like we need unit tests for this file and a couple other things:

image

File: Source/Core/ExperimentalFeatures.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+ * Flags to enable experimental features in CesiumJS.Stability and performance
```
Comment (ebogo1) [inline Source/Core/ExperimentalFeatures.js:2]:
```suggestion
 * Flags to enable experimental features in CesiumJS. Stability and performance
```

File: Source/Core/ExperimentalFeatures.js (hunk 1 above)
Comment (ebogo1) [reply Source/Core/ExperimentalFeatures.js:1]: looks good to me, thanks!

File: Specs/Scene/ModelExperimental/ModelExperimentalSceneGraphSpec.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import { ModelExperimentalSceneGraph } from "../../../Source/Cesium.js";
+
+describe("Scene/ModelExperimental/ModelExperimentalSceneGraph", function () {
```
[the remark this answers is not in the document]
Comment (ptrgags) [reply Specs/Scene/ModelExperimental/ModelExperimentalSceneGraphSpec.js:3]: Note that I added the node transformation fix we added in the unlit branch, see https://github.com/CesiumGS/cesium/pull/9703#discussion_r681882885

File: Source/Scene/ModelExperimental/buildDrawCommand.js (hunk 9 above)
Comment (ptrgags) [reply Source/Scene/ModelExperimental/buildDrawCommand.js:1]: I made a change to convert our pipeline stages into namespaces rather than functions. It's more typical, and it should give better test coverage.

File: Source/Scene/ModelExperimental/ModelExperimentalUtility.js [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+/**
+ * Create a function for reporting when a model fails to load
+ *
+ * @param {ModelExperimental} model The model to report about
+ * @param {String} type The type of object to report about
+ * @param {String} path The URI of the model file
+ * @return {Function} An error function that
+ */
+ModelExperimentalUtility.getFailedLoadFunction = function (model, type, path) {
```
[the remark this answers is not in the document]
Comment (sanjeetsuhag) [reply Source/Scene/ModelExperimental/ModelExperimentalUtility.js:24]: I see this as a way to keep `ModelExperimental` as clean and readable as possible, keeping things like this tucked away.

File: Source/Scene/ModelExperimental/ModelExperimental.js [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+/**
+ * A 3D model based on glTF, the runtime asset format for WebGL. This is
+ * a new architecture that is more decoupled than the older {@link Model}.
+ *
+ * This class is still experimental. glTF features that are core to 3D Tiles
+ * are supported, but other features such as animation are not yet supported.
+ *
+ * @constructor
+ *
+ * @param {Object} options Object with the following properties:
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/ModelExperimental.js:18]: @sanjeetsuhag I just noticed, this class has many more options that are not documented

File: Source/Scene/ModelExperimental/ModelExperimental.js [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+ * @param {Uint8Array} [options.gltf] A binary glTF buffer.
+ * @param {Resource|String} [options.basePath=''] The base path that paths in the glTF JSON are relative to.
+ *
+ * @private
+ * @experimental This feature is using part of the 3D Tiles spec that is not final and is subject to change without Cesium's standard deprecation policy.
+ */
+export default function ModelExperimental(options) {
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/ModelExperimental.js:25]: Another thing, @sanjeetsuhag are we intending to add `.fromGltf()` here or in another PR?

File: Source/Scene/ModelExperimental/ModelExperimental.js (hunk 13 above)
Comment (ptrgags) [reply Source/Scene/ModelExperimental/ModelExperimental.js:18]: Note that some options are set from `Gltf3DTileContent.update()` not the constructor

**Review by ptrgags — COMMENTED** [review]
@sanjeetsuhag noticed a couple more things as I was trying to adapt the IBL sandcastle for ModelExperimental

File: Source/Scene/ModelExperimental/ModelExperimental.js (hunk 14 above)
Comment (sanjeetsuhag) [reply Source/Scene/ModelExperimental/ModelExperimental.js:25]: I think we should focus on 3D Tiles only for now.

File: Specs/Scene/ModelExperimental/GeometryPipelineStageSpec.js [hunk 15]
```diff
@@ -0,0 +1,564 @@
...
+      "./Data/Models/GltfLoader/BuildingsMetadata/glTF/buildings-metadata.gltf";
+
+    var scene;
+    var gltfLoaders = [];
+
+    beforeAll(function () {
+      scene = createScene();
+    });
+
+    afterAll(function () {
```
Comment (ptrgags) [inline Specs/Scene/ModelExperimental/GeometryPipelineStageSpec.js:64]: @sanjeetsuhag add an afterEach() function that calls `ResourceCache.clearForSpecs()`, otherwise you'll sometimes get failures due to double-unloading things from the cache.

See `ResourceCacheSpec` for an example.

Comment (ptrgags) [reply Specs/Scene/ModelExperimental/GeometryPipelineStageSpec.js:64]: Er wait totally didn't see the afterEach() below this comment

Comment (ptrgags) [reply Specs/Scene/ModelExperimental/GeometryPipelineStageSpec.js:64]: But check in other files, `ModelExperimentalSpec` is one case that probably needs the afterEach(), perhaps others.

File: Source/Scene/ModelExperimental/ModelExperimentalUtility.js [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+};
+
+/**
+ * Create a bounding sphere from a primitive's POSITION attribute and model
+ * matrix.
+ *
+ * @param {ModelComponents.Primitive} primitive The primitive components.
+ * @param {Matrix4} modelMatrix The primitive's model matrix.
+ */
+ModelExperimentalUtility.createBoundingSphere = function (
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/ModelExperimentalUtility.js:86]: @sanjeetsuhag Testing out a model on another branch, I found a case where the bounding sphere isn't being computed correctly:
image


the bounding spheres should look like this: (old Model)
image

[Sandcastle with debug volumes](http://localhost:8080/Apps/Sandcastle/index.html#c=pVh7c9u4Ef8qqP6J1JEhEiRF0q/p2U5yaZ3GEzuZ6Z1uHIiERNQkwQFB2+qNv3sX4NvSXe5aj0ciFot9/PbBhR6pRI+cPTGJzlDOntAlK3mV4a+GNl1NIrO+FLmiPGdyNZmdrPJVzjdo+pf6IC4jljNcFiyqUirf5o9cijxjufpIi/K2KgohFYtn6NdVjtATz2PxhGnKpJpqAkKryV3CS7SW4qkEO2LBSpQLhcr6KGolI9aLRhnIxquJlqAtetFGPY55QP2Xz9foTPOsJolSRXm8WNQO4Uhki6Japzxa3NI8jmipUnZLsyJlV1TRxQOnkt4v7+lGMZkLkd+Th3u+TvGDeiaryUmtLRMxS40S0IDxAv4HMsxuubihT3nzibfp2hxe5YsFMm6/cgo90RIVUoCZJYtRVfJ8i/6RAIco35Rom969Qx8urlGtRmJ0J1p2tBOVROIpPzbSbXSErmCVChojcBGtK57GSCWslbcnDU1blLZcJdXaoNQwv5eiKhb6xBGcOGpOzLDRRUDX5ypH36KUoyOeF5W6oSpBiwI+F0oseEa3DCexREeiUpfVmkF0+m2gwRGD7Tc8kJR8QxshEU1TJArFBSDYYifQFtJOUsWMS2WRMMkjmqKEykzkPEKRYJsNjzggC9nFUvE0BzwZei/ENmUA5jueUoM7APhvFqkDuGnRHZtkKaMg4BBKWyN0sWl4Fw1vuQdQBmajoyO1K9gZ+AuPMStSsTt7BcYB7ACaT2CQ1JhkFJyikgE2pWhLhcV4rCVhaXEIQrDIAYvu6EONXgtlPAZNZ00LQ4Z43nSHv98iSNIyAZBqXHFbfdeWNe4jlxSsKjnNHVPsNiaO61vhkliu7/nLcN5Qie0ubRIurcD2XKel+o7tLYPAdhzihP4qnzVld23f29/RYzteSGybWEvXWdrLRqLt28T2/TC0g6Vv+a2ewF+6ru86oR06vjPU8113XE+fscNg6QQhmXdUD1T4QRBaoKtTQwIghi7xLDcIhmp+35sjMNxaho7lO8QnvuXVEjXZJh4hLgj2iWWFHdlzwKLACYjjhO5QE7kn31NlBUtClqFr2Z5nkU6m5YcQjtC1CXFs1+tUWXZoEzcMAD/AdqTquzEKwgDSwXZCxwu8Fjw78EILog5+um6fC7brAI9vWz5xRuCR34+RhS2IeGC5HqTXssk4IBLLhlh7rsmxjgjAOd4yJA4Ea+SM/QdgCwA2YnlW6DhBB1sAiW2DBeCL6wV94CA1XQsSgoSeP3LnOwECKz3i2m7gQoisBjVNdQPHAsUW1BBkX0+2oaog6S0vWHphr2hU6GfoZyjduSks/Wke9RPki/k0z5pKzBP55aQt+YTxbaJAAuhqRCfFq2niR0ZjeI/dcBUln0WaToF3jtqP1iIh+RZ6zNm+53gjRXbFtpKxssGbOJCRLmBbe+q6AIvl+FYdytqo3lnzJv5IleTPvfw7SfNSt9ISJ68MvBPv+DOL30lo5kZhbVstu5C14IFoEDoahgrJM674IysxjWMjoVH6UbMbf96najP9tR6BKpked8PEvKYNbD4eLtptnvOsym7AzvSW/4cdI5sEzV7M1tX2Fvrzhahy7dhXkVYZsChZMc3yMmtdMIKxBP93N2AUL80+hoafTzdVHuk3BpoarmaCQ8ikDyAjae93vTYiNQu8YH4SIkNK1KYPDsIoKQFiM3GOJ8gI4pvfFjRil0baZcd60p+X9UCn/wi20F8RoJLgjD7XRuJ14/KtHggYlhDYqpw39gLwVakgDDmjctZI7S3CDaja9CteKppHDKyUoMXCXueccQPqxrjQxtXExsVZlSpepLuL3Y3geTviNtHE+2Fsd16ZXYvveA73g1m9PRug0yTyyK4EK/EZyJDtU+I0BdceKHTK/yb7ERmwNxCmQjz80Dv225aOqp7mWzZtrJvXWucGWAjiKz9Geroinb4C+sPV23/efbj716wLSw1kNwj+2MyBl+NGN+x7J+ODBy8wcGTvTjFM8zsYn3Qam8KGAqPRQ1lPpErPpmKD9FSuEwvRooD7BtUlhfsI9IfP2gJDCMqV5zr9flA/sRzGzKYF4L2NBvmXk84iKJtHuF0ZI3rhGcvWTJYwxUFNPuQieoA5E4k1XLke6TplZWNSA3PLgY1H007ObOj7BVzmXqkB4Zpw9ekjggk4M6ADBpr25QN8UQXxhVlUz6RcDWBQQqRrqisqFlGlD+ItU29rGRe7D9BFJw1PfQk9ZKxGeKetgiwre6PnrfTe+q63ldUaGg9fsxuquz3k8jSH71kfi1dKWjJ0STDwUwfgUJ2RMGDslAx6KlTKV5pWQ01dMv6sBfxSv0QN00nP8tI6/9L6csCF1WQvUdqruz7RX3n1++lObOHuclEpJfKurleTL3DVMVfLuJJwrxpeVFP9egWMV5Ou8DdwEWH9qnMzSlj00F3/6z/z68Ghjf+jiMfH/0wpd8BCxoLL/4M50LrZhucs/lO2HDr10j6+DHvii8klLPTl7wle0YMsYlIK2WE4+nWl3moEwNdkPjkt1S5l562Sv/HM/MIC48cU44VicKeHplUu1hWERuGoLNtkO10Mj57G/BHx+OzA70MoSmlZws6mSs1gspqcny6Af++ovllCDn2CZpXSnWZL7PPrmogxPl3A8vDJrg90npwqXYLnPZCnai3i3fkwHqdKjtaaEp9ft3WCoDXVlXK6APo+55gCNPM7BzI3+dVE6jccuA/SYAWvTf1Mn+GZmOdSsaLesGEVU0WP1hAtID3qAj/eb/lzZHa+FMAM+2+MujcDpw8botiz0voAeli5f1DZntg9EIAwxA+WI4RhPQjBMG7/BQ)

Might be good to check `ModelUtility.computeBouningSphere` -- it's doing something more involved with the scene graph

Comment (sanjeetsuhag) [reply Source/Scene/ModelExperimental/ModelExperimentalUtility.js:86]: image

Fixed in the latest version.

File: Source/Scene/ModelExperimental/ModelExperimental.js (hunk 13 above)
Comment (sanjeetsuhag) [reply Source/Scene/ModelExperimental/ModelExperimental.js:21]: I am currently going with the approach of documenting what we're actually using in `ModelExperimental`. I wish we could remove the unused parameters, however, I can't because of `getModelClass`.

**Comment by sanjeetsuhag** [comment]
@ptrgags This is ready for review. The tests are passing locally for me, but I'll continue trying to figure out what's breaking CI.

File: Source/Scene/ModelExperimental/buildDrawCommand.js (hunk 9 above)
Comment (sanjeetsuhag) [reply Source/Scene/ModelExperimental/buildDrawCommand.js:1]: image

I got the coverage up. There are couple of things that would be easily tested with a `verifyRender` call, but this is not ready for that yet since the fragment shader is so primitive.

Comment (ptrgags) [reply Source/Scene/ModelExperimental/buildDrawCommand.js:1]: Yeah  that's fine. We should talk about when to add rendering tests soon (e.g. which branch to start introducing that)

File: Source/Scene/ModelExperimental/GeometryPipelineStage.js [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+  var variableName;
+  var varyingName;
+  var glslType = AttributeType.getGlslType(attributeType);
+
+  if (defined(semantic)) {
+    variableName = VertexAttributeSemantic.getVariableName(semantic, setIndex);
+    varyingName = "v_" + variableName;
+
+    shaderBuilder.addVarying(glslType, varyingName);
```
[the remark this answers is not in the document]
Comment (ptrgags) [reply Source/Scene/ModelExperimental/GeometryPipelineStage.js:70]: In Slack, we discussed this a bit, we want to do two things for now:

1. Always add varyings to the shader even for custom shaders
2. also generate a GLSL function to assign `v_customAttribute = a_customAttribute` and call it in the VS shader

eventually we need a find a way to disable some of this code if not actually used in the shader. But this will be handled in future PRs.

File: Source/Scene/ModelExperimental/GeometryPipelineStage.js [hunk 18]
```diff
@@ -0,0 +1,121 @@
+import defined from "../../Core/defined.js";
+import PrimitiveType from "../../Core/PrimitiveType.js";
+import AttributeType from "../AttributeType.js";
+import VertexAttributeSemantic from "../VertexAttributeSemantic.js";
+import GeometryVS from "../../Shaders/ModelExperimental/GeometryVS.js";
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/GeometryPipelineStage.js:5]: Rename to `GeometryStageVS` for consistency

File: Source/Scene/ModelExperimental/ModelExperimental.js [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+  destroyObject(this);
+};
+
+function initialize(model, options) {
+  var gltf = options.gltf;
+
+  var loaderOptions = {
+    baseResource: options.basePath,
+    releaseGltfJson: options.releaseGltfJson,
+    incrementallyLoadTextures: options.incrementallyLoadTextures,
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/ModelExperimental.js:197]: I don't see these options documented in the constructor.

File: Source/Scene/ModelExperimental/ModelExperimental.js [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+        this,
+        "model",
+        options.basePath
+      );
+    });
+}
+
+/**
+ *
+ * @param {Object}} options Object with the following properties:
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/ModelExperimental.js:237]:
```suggestion
 * @param {Object} options Object with the following properties:
```

File: Source/Scene/ModelExperimental/ModelExperimental.js [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+ * @param {Resource|String} options.url The url to the .gltf or .glb file.
+ * @param {Object} [options.basePath=''] The base path that paths in the glTF JSON are relative to.
+ * @param {Matrix4} [options.modelMatrix=Matrix4.IDENTITY] The 4x4 transformation matrix that transforms the model from model to world coordinates.
+ * @param {Boolean} [options.incrementallyLoadTextures] Determine if textures may continue to stream in after the model is loaded.
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/ModelExperimental.js:241]: Make sure these parameter lists are consistent between `fromGltf` and the constructor (where applicable)

File: Source/Scene/ModelExperimental/ModelExperimentalUtility.js [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+/**
+ * Model matrices in a model file (e.g. glTF) are typically in a different
+ * coordinate system, such as with y-up instead of z-up. This method adjusts
+ * the matrix so z is up, x is forward.
+ *
+ * @param {Matrix4} modelMatrix The original model matrix. This will be updated in place
+ * @param {Axis} upAxis The original up direction
+ * @param {Axis} forwardAxis The original forward direction
+ */
+ModelExperimentalUtility.correctModelMatrix = function (
```
[the remark this answers is not in the document]
Comment (ptrgags) [reply Source/Scene/ModelExperimental/ModelExperimentalUtility.js:112]: Eh probably not worth worrying about right now.

File: Source/Scene/ModelExperimental/ModelExperimentalUtility.js [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+    }
+  }
+};
+
+/**
+ * Create a bounding sphere from a primitive's POSITION attribute and model
+ * matrix.
+ *
+ * @param {ModelComponents.Primitive} primitive The primitive components.
+ * @param {Matrix4} modelMatrix The primitive's model matrix.
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/ModelExperimentalUtility.js:84]: Make sure these get marked as @private if they're not part of the public API

**Review by ptrgags — COMMENTED** [review]
@sanjeetsuhag looking good, just a few small tweaks.

I'll update my unlit branch and try out your changes

File: Source/Scene/ModelExperimental/GeometryPipelineStage.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (sanjeetsuhag) [reply Source/Scene/ModelExperimental/GeometryPipelineStage.js]: This has now been added. It's not a function, just a simple list of statements.

File: Source/Scene/ModelExperimental/buildDrawCommand.js [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+  meshPrimitiveRenderResources,
+  frameState
+) {
+  var shaderBuilder = meshPrimitiveRenderResources.shaderBuilder;
+  shaderBuilder.addVertexLines([ModelExperimentalVS]);
+  shaderBuilder.addFragmentLines([ModelExperimentalFS]);
+
+  var vertexArray = new VertexArray({
+    context: frameState.context,
+    indexBuffer: meshPrimitiveRenderResources.indices.buffer,
```
[the remark this answers is not in the document]
Comment (sanjeetsuhag) [reply Source/Scene/ModelExperimental/buildDrawCommand.js]: image
 Works now.

File: Specs/Scene/ModelExperimental/ModelExperimentalSpec.js [hunk 25]
```diff
@@ -0,0 +1,28 @@
+import {
+  GltfLoader,
+  Resource,
+  ModelExperimental,
+} from "../../../Source/Cesium.js";
+
+describe("Scene/ModelExperimental", function () {
```
[the remark this answers is not in the document]
Comment (sanjeetsuhag) [reply Specs/Scene/ModelExperimental/ModelExperimentalSpec.js]: I think once we have UNLIT, we'll be good to go.

**Review by ptrgags — COMMENTED** [review]
@sanjeetsuhag looks good aside from the mysterious Travis CI bug. Let's debug that together this afternoon

**Comment by ptrgags** [comment]
@sanjeetsuhag looks good now that we've fixed the unit tests! I'm merging now. Thanks!!
